### PR TITLE
Uploading the code base with Travis CI integrated in ROS 2

### PR DESCRIPTION
**Pull request by dejaniraai**
This PR comprises the code base of the main repository. This repository implements a state machine as prescribed in the Packaging Machine Language (PackML) standard, in simulation. This package also contains an RViz 2 plugin to be able to visualize the state of the state machine, the elapsed time in that state, and to control the triggering of state transitions of the machine through buttons. Finally, the package contains an example of the interface of the RViz2 plugin with a real PackML state machine running in a PLC, communicating its state and triggering transitions via OPCUA public tags. 

List of packages included in the repository:
* `packml_plugin`: RViz2 plugin for PackML state machine standard template visualization and control
* `packml_sm`: Simulator library in C++ ported to ROS 2 from the original PackML repository in ROS 1. Two types of state machines, with continuous Execute state and with timed Execute state.
* `packml_msgs`: Service type definitions for states, transitions and GUI control
* `packml_ros`: ROS 2 node in C++ to run the simulator library and communicate with the RViz2 plugin.

Extras:
* `packml_plc`: Example of a driver in Python to interface with a PackML state machine implemented in a Siemens PLC (with pre-configured OPCUA variable tags according to the PLCs configuration). Direct communication with the RViz2 plugin (receive states and send events to trigger transitions). 

A Travis CI pipeline has been added to test the code using a Docker container with ROS 2 dashing.

### CODE REVIEW

**Comment by dejaniraai** [comment]
Hi @gavanderhoorn, @ipa-nhg , @Levi-Armstrong, could you please help to review this pull request? I've tried to upload the code with a working Travis CI in place (I have tested with my own repository). Thank you for your help :)

**Comment by gavanderhoorn** [comment]
@dejaniraai: it may take me some time, but I'll try.

@Levi-Armstrong and @ipa-nhg: would be great if you could also take a look.

File: packml_plugin/include/packml_plugin/packml_widget.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  /** 
+  * @brief Node for the Widget or RViz plugin
+  */
+  std::shared_ptr<rclcpp::Node> nh_;
+
+
+  /**
+  * @brief Creates the GUI layout with an UI object as reference
+  */
+  Ui::PackmlPanel* ui_;
```
Comment (Levi-Armstrong) [inline packml_plugin/include/packml_plugin/packml_widget.h:63]: Could use a unique pointer here.

File: packml_plugin/include/packml_plugin/packml_plugin.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  /**
+  * @brief Constructor of the class
+  */
+  PackmlPlugin(QWidget* parent = 0);
+
+
+  /**
+  * @brief Destructor of the class
+  */
+  virtual ~PackmlPlugin() {};
```
Comment (Levi-Armstrong) [inline packml_plugin/include/packml_plugin/packml_plugin.h:47]: Recommend removing virtual and add keyword `override` instead and could set equal to default `= default`.

File: packml_plugin/include/packml_plugin/packml_widget.h [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+  * @brief Constructor of the class that creates the GUI plugin in Rviz, inherits a Widget
+  */
+  PackmlWidget(QWidget* parent = 0);
+
+
+  /**
+  * @brief Destructor for the widget object
+  */
+  virtual ~PackmlWidget();
```
Comment (Levi-Armstrong) [inline packml_plugin/include/packml_plugin/packml_widget.h:51]: Recommend removing virtual and adding keyword override.

File: packml_plugin/src/packml_widget.cpp [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+
+PackmlWidget::PackmlWidget(QWidget* parent) : QWidget(parent)
+{
+  nh_ = rclcpp::Node::make_shared("uiPlugin");  // Creates a node
+  // UI setup
+  ui_ = new Ui::PackmlPanel;
+  ui_->setupUi(this);
+  QTimer *timer = new QTimer(this);
+  connect(timer, SIGNAL(timeout()), this, SLOT(statusRequest()));
+  timer->start(200);
```
Comment (Levi-Armstrong) [inline packml_plugin/src/packml_widget.cpp:36]: Who deletes timer?

File: packml_plugin/src/packml_widget.cpp [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+
+
+void PackmlWidget::statusRequest()
+{
+  auto update = std::make_shared<packml_msgs::srv::AllStatus::Request>();
+  update->command = true;
+  auto result_future = status_client_->async_send_request(update);
+  if (rclcpp::spin_until_future_complete(nh_, result_future) != rclcpp::executor::FutureReturnCode::SUCCESS) {
+    printf("Error in service call, check that the ROS node or PLC driver are up and running\n");
+  } else {
```
Comment (Levi-Armstrong) [inline packml_plugin/src/packml_widget.cpp:64]: Style guide.

File: packml_plugin/src/packml_widget.cpp [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+
+void PackmlWidget::updateButtonState(std::shared_ptr<packml_msgs::srv::AllStatus::Response> msg)
+{
+  disableAllButtons();
+  if (msg->stopped_state == true)
+  {
+    ui_->stopped_state->setStyleSheet("QLabel { color : red; }");
+    ui_->abort_button->setEnabled(true);
+    ui_->reset_button->setEnabled(true);
+  } else {
```
Comment (Levi-Armstrong) [inline packml_plugin/src/packml_widget.cpp:142]: Style Guide If, else brackets.

**Review by Levi-Armstrong — COMMENTED** [review]
These are cosmetic. Also it may be good to run clang format to address the ros style guide syntax. I have a format file in Tesseract that works well. I am about half way through and will take a look tomorrow.

**Comment by Briancbn** [comment]
@gavanderhoorn 
> `rosdep` is failing, because these are not correct:
> 
> https://github.com/dejaniraai/packml_ros2/blob/69020fd6fe44515891291887101205b18cf44522/packml_ros/package.xml#L17
> 
> `Qt5` is not the correct `rosdep` key for Qt. Same problem with the `exec_depend` and in `packml_sm`.
> 
> You'll need to determine which parts of Qt5 are needed, and add the correct keys to the `package.xml` for all packages.
> 
> See [here](https://github.com/ros/rosdistro/blob/8814ff30b1dd25ad5c6ed8bf686205ed7fb466e2/rosdep/base.yaml#L3075) for one of the Qt5 keys. Search for `libqt5` for the others.

Fixed in https://github.com/ros-industrial/packml_ros2/commit/efe3b95364f682ce5f6c474eb17c7eccfebabaa4#diff-9c4089582270fc88290883b4fa52ce1fR16

File: packml_plugin/include/packml_plugin/packml_widget.h (hunk 1 above)
Comment (Briancbn) [reply packml_plugin/include/packml_plugin/packml_widget.h:63]: Addressed in https://github.com/ros-industrial/packml_ros2/pull/2/commits/3c153847145310fca867ccce6de740feebf7f855

File: packml_plugin/include/packml_plugin/packml_plugin.h (hunk 2 above)
Comment (Briancbn) [reply packml_plugin/include/packml_plugin/packml_plugin.h:47]: Addressed in https://github.com/ros-industrial/packml_ros2/pull/2/commits/9eac2fc8053fae7e99283b23314dd6d867b05c4c

File: packml_plugin/include/packml_plugin/packml_widget.h (hunk 3 above)
Comment (Briancbn) [reply packml_plugin/include/packml_plugin/packml_widget.h:51]: Addressed in https://github.com/ros-industrial/packml_ros2/pull/2/commits/9eac2fc8053fae7e99283b23314dd6d867b05c4c

File: packml_plugin/src/packml_widget.cpp (hunk 4 above)
Comment (Briancbn) [reply packml_plugin/src/packml_widget.cpp:36]: @Levi-Armstrong I used `timerEvent` for this here in https://github.com/ros-industrial/packml_ros2/pull/2/commits/de9d8f837af241c3d71fbf76ababad5f3c3672ec.

**Comment by cboostjvisser** [comment]
Hello, 

Does this PR still needs merging? Or can this one be closed because of #2?

**Comment by Briancbn** [comment]
Thanks for the reminder. I will close this.

**Pull request closed without merging**